File: packages/sdk-java/src/main/java/org/melaya/HttpClient.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Internal HTTP client. Sends the API key ONLY as an {@code Authorization: Bearer}
 * header on every call — never in the URL query string, so the key cannot leak
 * into access logs, proxies, or referrer headers. TLS certificate and hostname
 * verification always use the JVM trust configuration.
 */
public class HttpClient {

    static final ObjectMapper MAPPER = new ObjectMapper();

    private final String apiKey;
    private final String baseUrl;
    private final java.net.http.HttpClient http;

    public HttpClient(String apiKey, String baseUrl) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;

        this.http = java.net.http.HttpClient.newBuilder().build();
    }

    /** Builds a full URL with the given query params.
     *
     * SECURITY: the API key is NOT injected here. It travels only in the
     * {@code Authorization: Bearer} header (set on every request below), so it
     * can never leak into access logs, proxies, or referrer headers. */
    String buildUrl(String path, Map<String, Object> query) {
        StringBuilder sb = new StringBuilder(baseUrl);
        if (!path.startsWith("/")) sb.append("/");
        sb.append(path);
        Map<String, Object> params = new LinkedHashMap<>();
        if (query != null) {
            for (Map.Entry<String, Object> e : query.entrySet()) {
                if (e.getValue() != null) params.put(e.getKey(), e.getValue());
            }
        }
        boolean first = true;
        for (Map.Entry<String, Object> e : params.entrySet()) {
            sb.append(first ? "?" : "&");
            sb.append(URLEncoder.encode(e.getKey(), StandardCharsets.UTF_8))
              .append("=")
              .append(URLEncoder.encode(String.valueOf(e.getValue()), StandardCharsets.UTF_8));
            first = false;
        }
        return sb.toString();
    }

    public JsonNode get(String path, Map<String, Object> query) {
        String url = buildUrl(path, query);
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .GET()
                .build();
        return execute(req);
    }

    public JsonNode post(String path, Object body) {
        String url = buildUrl(path, null);
        String json;
        try {
            json = body == null ? "" : MAPPER.writeValueAsString(body);
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize request body", e);
        }
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
        return execute(req);
    }

    public JsonNode put(String path, Object body) {
        String url = buildUrl(path, null);
        String json;
        try {
            json = body == null ? "" : MAPPER.writeValueAsString(body);
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize request body", e);
        }
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .PUT(HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
        return execute(req);
    }

    public JsonNode patch(String path, Object body) {
        String url = buildUrl(path, null);
        String json;
        try {
            json = body == null ? "" : MAPPER.writeValueAsString(body);
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize request body", e);
        }
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .method("PATCH", HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
        return execute(req);
    }

    public JsonNode delete(String path, Map<String, Object> query) {
        String url = buildUrl(path, query);
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .DELETE()
                .build();
        return execute(req);
    }

    /** Per-request timeout in milliseconds (default 30 s). */
    static final long REQUEST_TIMEOUT_MS = 30_000;

    /**
     * Max retries for idempotent GET requests on network error / 429 / 5xx.
     * Non-GET methods are never retried.
     */
    private static final int MAX_GET_RETRIES = 2;

    /** Base backoff for retries (ms). Actual delay = base * 2^attempt + jitter(0..200 ms). */
    private static final long RETRY_BASE_MS = 500;

    private JsonNode execute(HttpRequest req) {
        boolean isGet = req.method().equalsIgnoreCase("GET");
        int maxRetries = isGet ? MAX_GET_RETRIES : 0;
        IOException lastIoEx = null;

        // Rebuild the request with a per-request timeout.
        req = HttpRequest.newBuilder(req, (k, v) -> true)
                .timeout(Duration.ofMillis(REQUEST_TIMEOUT_MS))
                .build();

        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
                int status = resp.statusCode();
                String rawBody = resp.body();
                JsonNode data = null;
                if (rawBody != null && !rawBody.isBlank()) {
                    try {
                        data = MAPPER.readTree(rawBody);
                    } catch (Exception e) {
                        // not JSON; wrap as text node
                        data = MAPPER.getNodeFactory().textNode(rawBody);
                    }
                }
                // Retry GET on 429 or 5xx (bounded, with Retry-After support)
                if (isGet && (status == 429 || status >= 500) && attempt < maxRetries) {
                    long delay = backoffDelay(attempt);
                    String retryAfter = resp.headers().firstValue("Retry-After").orElse(null);
                    if (retryAfter != null) {
                        try { delay = Long.parseLong(retryAfter.trim()) * 1000L; } catch (NumberFormatException ignored) {}
                    }
                    sleep(delay, req);
                    continue;
                }
                if (status >= 400) {
                    String code = null;
                    String message = null;
                    if (data != null && data.isObject()) {
                        if (data.has("error")) code = data.get("error").asText(null);
                        if (data.has("message")) message = data.get("message").asText(null);
                    }
                    // Never surface the raw auth header in exceptions
                    String detail = code != null ? " (" + code + ")" : (message != null ? " (" + message + ")" : "");
                    throw new MelayaException(
                            "Melaya API " + status + detail,
                            status, code, data);
                }
                // Check ok: false envelope
                if (data != null && data.isObject()) {
                    JsonNode okNode = data.get("ok");
                    if (okNode != null && okNode.isBoolean() && !okNode.asBoolean()) {
                        String code = data.has("error") ? data.get("error").asText(null) : null;
                        throw new MelayaException(
                                "Melaya API request failed" + (code != null ? ": " + code : ""),
                                status, code, data);
                    }
                }
                return data;
            } catch (MelayaException me) {
                throw me; // never retry business-logic errors
            } catch (IOException e) {
                lastIoEx = e;
                if (isGet && attempt < maxRetries) {
                    sleep(backoffDelay(attempt), req);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException("HTTP request interrupted: " + req.uri(), e);
            }
        }
        throw new RuntimeException(
                "HTTP request failed after " + maxRetries + " retries: " + req.uri(), lastIoEx);
    }

    /** Exponential backoff with up to 200 ms of random jitter. */
    private static long backoffDelay(int attempt) {
        long base = RETRY_BASE_MS * (1L << attempt); // 500, 1000
        long jitter = ThreadLocalRandom.current().nextLong(0, 200);
        return base + jitter;
    }

    private static void sleep(long ms, HttpRequest req) {
        try { Thread.sleep(ms); } catch (InterruptedException ie) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("HTTP retry interrupted: " + req.uri(), ie);
        }
    }

    public String getApiKey() { return apiKey; }
    public String getBaseUrl() { return baseUrl; }
    public java.net.http.HttpClient getHttpImpl() { return http; }
}

File: packages/sdk-kotlin/src/main/kotlin/org/melaya/HttpClient.kt
package org.melaya

import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject

private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

/**
 * OkHttp interceptor that retries idempotent GET requests on network error,
 * HTTP 429 (rate-limited), or 5xx (server error).
 *
 * Rules:
 *   - Only retries GET requests (POST/PUT/PATCH/DELETE are never retried).
 *   - Maximum 2 retries (3 total attempts).
 *   - Exponential back-off with ±20% jitter: ~1s, ~2s.
 *   - Respects the `Retry-After` header on 429 (treated as seconds, capped at 30s).
 *   - Secrets (API key) are never surfaced in exceptions or logs.
 */
internal class RetryInterceptor(private val maxRetries: Int = 2) : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        // Only retry idempotent GETs
        if (request.method != "GET") return chain.proceed(request)

        var attempt = 0
        var response = chain.proceed(request)
        while (attempt < maxRetries && shouldRetry(response.code)) {
            response.close()
            Thread.sleep(retryDelayMs(response, attempt))
            attempt++
            response = chain.proceed(request)
        }
        return response
    }

    private fun shouldRetry(code: Int) = code == 429 || code in 500..599

    private fun retryDelayMs(response: Response, attempt: Int): Long {
        val retryAfter = response.header("Retry-After")?.toLongOrNull()
        // Honor Retry-After but cap it at 30s so a hostile/buggy header can't stall the client.
        if (retryAfter != null) return retryAfter.coerceIn(0L, 30L) * 1_000L
        val base = (1L shl attempt) * 1_000L   // 1s, 2s
        val jitter = (base * 0.2 * Math.random()).toLong()
        return base + jitter
    }
}

/**
 * Internal HTTP client. Sends the API key ONLY as an `Authorization: Bearer`
 * header on every call — never in the URL query string, so the key cannot leak
 * into access logs, proxies, or referrer headers.
 *
 * @param requestTimeoutMs Per-request call timeout in ms (default 30 000). Applied via
 *   OkHttp's per-call [OkHttpClient.newBuilder] override so it does not affect the base
 *   client's connection/read/write timeouts.
 */
internal class HttpClient(
    private val apiKey: String,
    private val baseUrl: String,
    internal val okHttp: OkHttpClient,
    private val requestTimeoutMs: Long = 30_000L,
) {

    // ── URL helpers ──────────────────────────────────────────────────────────

    private fun buildUrl(path: String, query: Map<String, Any?> = emptyMap()): String {
        val base = baseUrl.trimEnd('/')
        val sb = StringBuilder("$base/${ path.trimStart('/') }")
        var first = true
        for ((k, v) in query) {
            if (v != null) {
                sb.append(if (first) "?" else "&").append("${encode(k)}=${encode(v.toString())}")
                first = false
            }
        }
        return sb.toString()
    }

    private fun encode(s: String) = java.net.URLEncoder.encode(s, "UTF-8")

    // ── Verb helpers ─────────────────────────────────────────────────────────

    fun get(path: String, query: Map<String, Any?> = emptyMap()): Any? {
        val req = Request.Builder()
            .url(buildUrl(path, query))
            .header("Authorization", "Bearer $apiKey")
            .get()
            .build()
        return execute(req)
    }

    private fun toRequestBody(body: Any?) = when (body) {
        null -> "{}".toRequestBody(JSON_MEDIA_TYPE)
        is JSONObject -> body.toString().toRequestBody(JSON_MEDIA_TYPE)
        is Map<*, *> -> JSONObject(body).toString().toRequestBody(JSON_MEDIA_TYPE)
        is List<*> -> org.json.JSONArray(body).toString().toRequestBody(JSON_MEDIA_TYPE)
        else -> body.toString().toRequestBody(JSON_MEDIA_TYPE)
    }

    fun post(path: String, body: Any? = null): Any? {
        val jsonBody = toRequestBody(body)
        val req = Request.Builder()
            .url(buildUrl(path))
            .header("Authorization", "Bearer $apiKey")
            .header("Content-Type", "application/json")
            .post(jsonBody)
            .build()
        return execute(req)
    }

    fun put(path: String, body: Any? = null): Any? {
        val jsonBody = toRequestBody(body)
        val req = Request.Builder()
            .url(buildUrl(path))
            .header("Authorization", "Bearer $apiKey")
            .header("Content-Type", "application/json")
            .put(jsonBody)
            .build()
        return execute(req)
    }

    fun patch(path: String, body: Any? = null): Any? {
        val jsonBody = toRequestBody(body)
        val req = Request.Builder()
            .url(buildUrl(path))
            .header("Authorization", "Bearer $apiKey")
            .header("Content-Type", "application/json")
            .patch(jsonBody)
            .build()
        return execute(req)
    }

    fun delete(path: String, query: Map<String, Any?> = emptyMap(), body: Any? = null): Any? {
        val reqBuilder = Request.Builder()
            .url(buildUrl(path, query))
            .header("Authorization", "Bearer $apiKey")
        val req = if (body != null) {
            reqBuilder.delete(toRequestBody(body)).build()
        } else {
            reqBuilder.delete().build()
        }
        return execute(req)
    }

    // ── Response parsing + envelope unwrap ──────────────────────────────────

    private fun execute(req: Request): Any? {
        // Apply a per-request call timeout without modifying the shared OkHttpClient.
        val callClient = if (requestTimeoutMs > 0) {
            okHttp.newBuilder()
                .callTimeout(requestTimeoutMs, java.util.concurrent.TimeUnit.MILLISECONDS)
                .build()
        } else {
            okHttp
        }
        val resp = callClient.newCall(req).execute()
        val text = resp.body?.string() ?: ""
        val status = resp.code

        val data: Any? = try {
            if (text.isBlank()) null
            else if (text.trimStart().startsWith("[")) JSONArray(text)
            else JSONObject(text)
        } catch (_: Exception) {
            text
        }

        if (status >= 400) {
            val code = (data as? JSONObject)?.optString("error", null)
            throw MelayaException(
                "Melaya API $status" + if (code != null) " ($code)" else "",
                status, code, data
            )
        }

        if (data is JSONObject && data.optBoolean("ok", true) == false) {
            val code = data.optString("error", null)
            throw MelayaException(
                "Melaya API request failed" + if (code != null) ": $code" else "",
                status, code, data
            )
        }

        return data
    }
}

// ── Convenience extension helpers used by API classes ──────────────────────

internal fun Any?.asObject(): JSONObject = this as? JSONObject
    ?: throw MelayaException("Unexpected response type: ${this?.javaClass?.simpleName}", 0)

internal fun Any?.asArray(): JSONArray = this as? JSONArray
    ?: throw MelayaException("Unexpected array response type: ${this?.javaClass?.simpleName}", 0)

internal fun JSONObject.getObject(key: String): JSONObject = optJSONObject(key)
    ?: throw MelayaException("Missing key '$key' in response", 0)

internal fun JSONObject.getArray(key: String): JSONArray = optJSONArray(key)
    ?: JSONArray()

internal fun JSONArray.toJsonObjects(): List<JSONObject> =
    (0 until length()).map { getJSONObject(it) }

internal fun JSONArray.toAnyList(): List<Any?> =
    (0 until length()).map { get(it) }
